### DISCUSSION TITLE
Split interfaces to different files

### DISCUSSION THREAD

**Discussion by sharpzilla** [christianhelle/refitter#385] [Q&A]
Hello, is there a way to generate file per tag when I'm generating clients with options "multipleInterfaces": "ByTag"?

Also is there a way to disable generation of models from open api specification? I have this structure of my solution: Api, Api.Client, Api.Models. Api.Models is shared project with projects: Api and Api.Client and it contains requets and response models and enums. Basically I don't need generated models from refitter I'm fine with my existed models from Api.Models.

**Comment by christianhelle** [comment]
@sharpzilla thanks for taking the time to bring this up

As of right now, there is no way to generate multiple files. This is something I am working on. 

To begin with, I want to enable an option to split the generated types into individual files. To begin with, that would be the following:
- Endpoint interfaces
- IServiceCollection extensions
- Models
 
Unfortunately, I piggyback entirely on NSwag for generating the models, and as far as I know, they only support generating a single file for models

I haven't thought about whether this would be the default, or should be enabled using some sort of option like `--multiple-files` or something like that. The way I use Refitter personally, and within the teams I work with is for generating a Client SDK for an API as part of an API's CI/CD pipeline, and the reason everything is just generated into a single file is because I'm only interested in the types generated, and not so much the layout of the file itself. I never really touch the generated file, and it's ignored from source control

**Reply by sharpzilla** [reply]
Would you be kindly to explain how you work with endpoints which return FileContentResult and "application/octet-stream" aka download that file. In our tasks most of the time we need extract byte array from HttpContent and get Filename from Headers and that process hided inside client from users of that client. With manual created client there is an option in refitter to use custom serializer. But in auto generated client for open api specification with response type application/octet-stream type string format binary will be generated FileResponse with dictionary collection of Headers and Stream property. So the user of that client should manually extract byte array and file name.

Also could you share your experience how you work with api clients with multiple controllers. I mean for me, it is comfortable to add api.client nuget reference in project and inject my top level API client interface which holds methods splitted by controllers hided in properties (i.e. `public IControlerA ControllerA {get;}`). You answered you generate clients using CI/CD is there a way to create client which i described in CI/CD?

**Comment by berkslv** [comment]
Hello @christianhelle, I realy like this project and want to contribute about this issue. I investigate the nswag generator toolchain and there is a option for generation only models and not client code ([Source](https://devblogs.microsoft.com/ise/maintain-api-client-with-nswag-model-gen/)). Maybe we create some sort of option like `--model-output` to create seperate models. 

This seperation is realy neccessary for me because we use clean architecture in our project and we want to store this generated client in our Infrastructure and models in Domain layer. But without seperation of client and models we force to use this generated file in our core (Business) layer but clean architecture recommends to store any external connection in Infra layer therefore we violated this clean architecture approach. For example, if you use database in your application, you put DbContext class in your Infra layer and put Entites class in Domain layer and there is genertor for this use case like `dbcontext scaffold`.

Are you recommend me someway to implement this, I realy want to contribute this great project?

**Reply by berkslv** [reply]
I opened a PR for this feature and added `--contract-only` option to application to create seperate client and contracts. 

https://github.com/christianhelle/refitter/pull/436

**Reply by christianhelle** [reply]
First, thank you for your interest and intent to contribute @berkslv

I like the `--contract-only` option you suggested and I'll take a look at your pull request

I'm sorry for the late reply, I'm currently on vacation and "trying" to be offline.

**Reply by berkslv** [reply]
Thanks for your efforts @christianhelle,

If there is a problem with PR, I would like to take care of it.

Have a nice holiday.
